### PR TITLE
[feature] support stable diffusion inference

### PR DESCRIPTION
**Pull request by liuwenran**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: tests/test_models/test_editors/test_ddim/test_ddim_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    with pytest.raises(Exception):
+        ddim.step(modelout, 980, sample)
+
+    ddim.set_timesteps(10)
+    result = ddim.step(
+        modelout, 980, sample, eta=1, use_clipped_model_output=True)
+    assert result['prev_sample'].shape == (1, 4, 32, 32)
+
+
+if __name__ == '__main__':
```
Comment (LeoXing1996) [inline tests/test_models/test_editors/test_ddim/test_ddim_scheduler.py:51]: I think we should remove `if __name__ == '__main__'` from all unit tests.

File: mmedit/models/editors/stable_diffusion/stable_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+                Pre-generated noisy latents, defaults to None.
+
+        Return:
+            latents (torch.Tensor): prepared latents.
+        """
+        shape = (batch_size, num_channels_latents,
+                 height // self.vae_scale_factor,
+                 width // self.vae_scale_factor)
+        if latents is None:
+            latents = torch.randn(
```
Comment (LeoXing1996) [inline mmedit/models/editors/stable_diffusion/stable_diffusion.py:480]: Will `torch.randn` use random seed set in `set_random_seed` if `generator` is None?

File: configs/stable_diffusion/base_config_denoisingunet.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+unet = dict(
+    type='DenoisingUnet',
+    image_size=512,
+    base_channels=320,
+    channels_cfg=[1, 2, 4, 4],
+    unet_type='stable',
+    act_cfg=dict(type='silu', inplace=False),
```
Comment (LeoXing1996) [inline configs/stable_diffusion/base_config_denoisingunet.py:7]: Seems only `type` is used. I think we can remove `inplace`.

File: configs/stable_diffusion/base_config_denoisingunet.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    config='config.json',
+    pytorch_model='pytorch_model.bin')
+
+feature_extractor_path = dict(
+    subdir_name='feature_extractor', config='preprocessor_config.json')
+
+safety_checker_path = dict(
+    subdir_name='safety_checker',
+    config='config.json',
+    pytorch_model='pytorch_model.bin')
```
Comment (LeoXing1996) [inline configs/stable_diffusion/base_config_denoisingunet.py:68]: These modules are initialized in `transformers`' style. I think we can provide some wrappers for these modules to implement openmmlab styles initialization. (This comment can be ignored if we do not have enough time.)

File: mmedit/models/editors/ddpm/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                 downscale_freq_shift: float = 0):
+        super().__init__()
+        self.num_channels = num_channels
+        self.flip_sin_to_cos = flip_sin_to_cos
+        self.downscale_freq_shift = downscale_freq_shift
+
+    def forward(self, timesteps):
+        """forward with timesteps."""
+
+        t_emb = get_timestep_embedding(
```
[the remark this answers is not in the document]
Comment (liuwenran) [reply mmedit/models/editors/ddpm/embeddings.py]: good idea, resolved.

File: mmedit/models/editors/stable_diffusion/stable_diffusion.py [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+from mmedit.registry import DIFFUSION_SCHEDULERS, MODELS
+from .clip_wrapper import load_clip_submodels
+from .vae import AutoencoderKL
+
+logger = MMLogger.get_current_instance()
+
+
+@MODELS.register_module('sd')
+@MODELS.register_module()
+class StableDiffusion(nn.Module):
```
[the remark this answers is not in the document]
Comment (liuwenran) [reply mmedit/models/editors/stable_diffusion/stable_diffusion.py:19]: resolved.

File: mmedit/models/editors/stable_diffusion/stable_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+    """class to run stable diffsuion pipeline.
+
+    Args:
+        diffusion_scheduler(dict):
+            Diffusion scheduler config.
```
[the remark this answers is not in the document]
Comment (liuwenran) [reply mmedit/models/editors/stable_diffusion/stable_diffusion.py:24]: resolved.

File: mmedit/models/editors/ddpm/embeddings.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                 in_channels: int,
+                 time_embed_dim: int,
+                 act_fn: str = 'silu',
+                 out_dim: int = None):
+        super().__init__()
+
+        self.linear_1 = nn.Linear(in_channels, time_embed_dim)
+        self.act = None
+        if act_fn == 'silu':
+            self.act = nn.SiLU()
```
[the remark this answers is not in the document]
Comment (liuwenran) [reply mmedit/models/editors/ddpm/embeddings.py:25]: check added.

File: mmedit/models/editors/stable_diffusion/stable_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+    def load_pretrained_ckpt(self, pretrained_ckpt_path):
+        """load pretrained ckpt for each submodel."""
+        if 'unet' in pretrained_ckpt_path.keys() and \
+                pretrained_ckpt_path['unet'] is not None:
+            state_dict = torch.load(
+                pretrained_ckpt_path['unet'], map_location='cpu')
+            self.unet.load_state_dict(state_dict, strict=True)
+
+        if 'vae' in pretrained_ckpt_path.keys() and \
+                pretrained_ckpt_path['unet'] is not None:
```
[the remark this answers is not in the document]
Comment (liuwenran) [reply mmedit/models/editors/stable_diffusion/stable_diffusion.py:72]: sorry, its a typo. resolved.

File: configs/stable_diffusion/base_config_denoisingunet.py (hunk 3 above)
Comment (liuwenran) [reply configs/stable_diffusion/base_config_denoisingunet.py]: resolved.

File: configs/stable_diffusion/base_config_denoisingunet.py (hunk 4 above)
Comment (liuwenran) [reply configs/stable_diffusion/base_config_denoisingunet.py]: yesyes, we should use openclip in the future.

File: mmedit/models/editors/stable_diffusion/stable_diffusion.py [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+        self.scheduler.order = 1
+        self.scheduler.init_noise_sigma = 1.0
+
+        self.unet_sample_size = unet_sample_size
+        self.unet = MODELS.build(unet) if isinstance(unet, dict) else unet
+
+        self.vae = AutoencoderKL(**vae) if isinstance(vae, dict) else vae
+        self.vae_scale_factor = 2**(len(self.vae.block_out_channels) - 1)
+
+        self.load_pretrained_ckpt(pretrained_ckpt_path)
```
[the remark this answers is not in the document]
Comment (liuwenran) [reply mmedit/models/editors/stable_diffusion/stable_diffusion.py]: good idea. resolved.

File: mmedit/models/editors/stable_diffusion/stable_diffusion.py [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+                class instance itsself.
+        """
+        if torch_device is None:
+            return self
+
+        for name in self.submodels:
+            module = getattr(self, name)
+            if isinstance(module, torch.nn.Module):
+                module.to(torch_device)
+        self.execution_device = torch.device(torch_device)
```
[the remark this answers is not in the document]
Comment (liuwenran) [reply mmedit/models/editors/stable_diffusion/stable_diffusion.py:98]: renamed.

File: mmedit/models/editors/stable_diffusion/stable_diffusion.py (hunk 2 above)
Comment (liuwenran) [reply mmedit/models/editors/stable_diffusion/stable_diffusion.py:479]: yes, set_random_seed will make seed be set globally.

File: tests/test_models/test_editors/test_ddim/test_ddim_scheduler.py (hunk 1 above)
Comment (liuwenran) [reply tests/test_models/test_editors/test_ddim/test_ddim_scheduler.py:51]: resolved.

File: mmedit/models/editors/stable_diffusion/clip_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+import os
+
+import torch
+import torch.nn as nn
+from mmengine.logging import MMLogger
+from transformers import CLIPConfig, CLIPVisionModel, PreTrainedModel
+from transformers.models.clip.feature_extraction_clip import \
+    CLIPFeatureExtractor  # noqa
+from transformers.models.clip.modeling_clip import CLIPTextModel
+from transformers.models.clip.tokenization_clip import CLIPTokenizer
```
Comment (LeoXing1996) [inline mmedit/models/editors/stable_diffusion/clip_wrapper.py:11]: We should use `try_import` here since `transformers` is an optional package.

e.g.,
```python
from mmediting import try_import
transformers = try_import('transformers')

if transformers is None:
        from transformers import CLIPConfig, CLIPVisionModel, PreTrainedModel
    from transformers.models.clip.feature_extraction_clip import \
        CLIPFeatureExtractor  # noqa
    from transformers.models.clip.modeling_clip import CLIPTextModel
    from transformers.models.clip.tokenization_clip import CLIPTokenizer

    class StableDiffusionSafetyChecker(PreTrainedModel):
        ...

    def load_clip_submodels(init_cfg, submodules, requires_safety_checker):
        ...

else:
    def load_clip_submodels(*args, **kwargs):
        raise ImportError('Please install tranformers via '
                          '\'pip install transformers\'')

```

Comment (LeoXing1996) [reply mmedit/models/editors/stable_diffusion/clip_wrapper.py:11]: Then we can use the following unit test to cover these modified lines:
```python
import pytest
import sys


def test_transformers_is_none():
    sys.modules['transformers'] = None
    from new_clip_wrapper import load_clip_submodels

    with pytest.raises(ImportError):
        load_clip_submodels('just a test')


def test_wrapper():
    ...

```

Comment (LeoXing1996) [reply mmedit/models/editors/stable_diffusion/clip_wrapper.py:11]: Then we can use the following unit test to cover these modified lines:
```python
import pytest
import sys


def test_transformers_is_none():
    sys.modules['transformers'] = None
    from new_clip_wrapper import load_clip_submodels

    with pytest.raises(ImportError):
        load_clip_submodels('just a test')


def test_wrapper():
    ...

```

Comment (liuwenran) [reply mmedit/models/editors/stable_diffusion/clip_wrapper.py]: resolved

**Review by LeoXing1996 — APPROVED** [review]
Good work! This PR looks good to me. 🚀
